modules/core/arrow-data/src/main/kotlin/arrow/data/EitherT.kt

```diff
@@ -21,66 +21,83 @@ data class EitherT<F, A, B>(val value: Kind<F, Either<A, B>>) : EitherTOf<F, A,
 
     companion object {
 
-        inline operator fun <reified F, A, B> invoke(value: Kind<F, Either<A, B>>): EitherT<F, A, B> = EitherT(value)
+        inline operator fun <F, A, B> invoke(value: Kind<F, Either<A, B>>): EitherT<F, A, B> = EitherT(value)
 
         fun <F, A, B> pure(b: B, MF: Applicative<F>): EitherT<F, A, B> = right(b, MF)
 
-        fun <F, L, A, B> tailRecM(a: A, f: (A) -> EitherTOf<F, L, Either<A, B>>, MF: Monad<F>): EitherT<F, L, B> =
-                EitherT(MF.tailRecM(a, {
-                    MF.map(f(it).fix().value) { recursionControl ->
-                        when (recursionControl) {
-                            is Either.Left<L, Either<A, B>> -> Right(Left(recursionControl.a))
-                            is Either.Right<L, Either<A, B>> -> {
-                                val b: Either<A, B> = recursionControl.b
-                                when (b) {
-                                    is Either.Left<A, B> -> Left(b.a)
-                                    is Either.Right<A, B> -> Right(Right(b.b))
-                                }
+        fun <F, L, A, B> tailRecM(a: A, f: (A) -> EitherTOf<F, L, Either<A, B>>, MF: Monad<F>): EitherT<F, L, B> = MF.run {
```

Comment:
Should functions like this depending on nested constrains be defined outside of companions such as that what now is:
```kotlin
fun <F, L, A, B> tailRecM(a: A, f: (A) -> EitherTOf<F, L, Either<A, B>>, MF: Monad<F>): EitherT<F, L, B>
```
becomes
```
fun <F, L, A, B> Monad<F>.tailRecM(a: A, f: (A) -> EitherTOf<F, L, Either<A, B>>): EitherT<F, L, B>
```

Reply:
Can't, because they're only available in the scope of EitherT. You'd have to do something like

```
val et = getEitherT()
et.run { MF.railrecM(...) }
```

to be able to use them. Typeclassless simply doesn't work with methods inside classes.